File: go/gherkin-generate-tokens/gherkin-generate-tokens.go
package main

import (
	"fmt"
	gherkin "github.com/cucumber/gherkin/go/v33"
	"io"
	"os"
	"strings"
)

func main() {

	var readers []io.Reader
	if len(os.Args) <= 1 {
		readers = append(readers, os.Stdin)
	} else {
		for i := range os.Args[1:] {
			file, err := os.Open(os.Args[i+1])
			if err != nil {
				fmt.Fprintf(os.Stderr, "Error: %s\n", err)
				os.Exit(1)
				return
			}
			defer file.Close()
			readers = append(readers, file)
		}
	}

	for i := range readers {
		err := GenerateTokens(readers[i], os.Stdout)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error: %s\n", err)
			os.Exit(1)
			return
		}
	}
}

type tokenGenerator struct {
	out io.Writer
}

func FormatToken(token *gherkin.Token) string {
	if token.IsEOF() {
		return "EOF"
	}
	var items []string
	for i := range token.Items {
		items = append(items, token.Items[i].String())
	}
	var keyword string
	if token.Keyword == "" {
		keyword = token.Keyword
	} else if token.KeywordType == "" {
		keyword = fmt.Sprintf("()%s", token.Keyword)
	} else {
		keyword = fmt.Sprintf("(%s)%s", token.KeywordType, token.Keyword)
	}
	var text string
	if token.Type.Name() == "Empty" {
		text = ""
	} else {
		text = token.Text
	}

	return fmt.Sprintf("(%d:%d)%s:%s/%s/%s",
		token.Location.Line,
		token.Location.Column,
		token.Type.Name(),
		keyword,
		text,
		strings.Join(items, ","),
	)
}

func (t *tokenGenerator) Build(tok *gherkin.Token) (bool, error) {
	fmt.Fprintln(t.out, FormatToken(tok))
	return true, nil
}
func (t *tokenGenerator) StartRule(r gherkin.RuleType) (bool, error) {
	return true, nil
}
func (t *tokenGenerator) EndRule(r gherkin.RuleType) (bool, error) {
	return true, nil
}
func (t *tokenGenerator) Reset() {
}

func GenerateTokens(in io.Reader, out io.Writer) error {

	builder := &tokenGenerator{out}
	parser := gherkin.NewParser(builder)
	parser.StopAtFirstError(true)
	matcher := gherkin.NewMatcher(gherkin.DialectsBuiltin())

	scanner := gherkin.NewScanner(in)

	return parser.Parse(scanner, matcher)
}

File: dart/bin/gherkin_generate_tokens.dart
import 'dart:io';

import 'package:cucumber_gherkin/src/language/dialects_builtin.g.dart';
import 'package:cucumber_gherkin/src/gherkin_token_matcher.dart';
import 'package:cucumber_gherkin/src/language/string_token_scanner.dart';
import 'package:cucumber_gherkin/src/parser/token.dart';
import 'package:cucumber_gherkin/src/parser/builder.dart';
import 'package:cucumber_gherkin/src/parser/parser.g.dart';

void main(List<String> args) {
  final matcher = GherkinTokenMatcher(builtinDialects);

  for (final path in args) {
    final builder = _TokenFormatterBuilder();
    Parser<void>(
      builder,
    ).parse(StringTokenScanner(File(path).readAsStringSync()), matcher);
    stdout.write(builder.tokensText);
  }
}

class _TokenFormatterBuilder implements Builder<void> {
  final StringBuffer _buffer = StringBuffer();

  String get tokensText => _buffer.toString();

  @override
  void build(Token token) => _buffer.writeln(_formatToken(token));

  @override
  void endRule(RuleType ruleType) {}

  @override
  void startRule(RuleType ruleType) {}

  @override
  void reset() {}

  @override
  void get result {}
}

String _formatToken(Token token) {
  if (token.isEof) {
    return 'EOF';
  }
  final matchedItems = token.matchedItems
      .map((element) => '${element.column}:${element.text}')
      .join(',');
  final matchedType = _wireName(token.matchedType);
  final matchedKeyword = token.matchedKeyword;
  final matchedKeywordType =
      matchedKeyword.isEmpty
          ? ''
          : token.matchedKeywordType == null
          ? '()'
          : '(${token.matchedKeywordType!.value})';
  return '(${token.location.line}:${token.location.column})'
      '$matchedType:$matchedKeywordType$matchedKeyword/${token.matchedText}/$matchedItems';
}

String _wireName(TokenType tokenType) {
  final name = tokenType.name;
  if (name == 'eof') {
    return 'EOF';
  }
  return '${name[0].toUpperCase()}${name.substring(1)}';
}
